subprojects/docs/src/docsTest/java/org/gradle/docs/samples/IntegrationTestSamplesExecutor.java

```diff
@@ -62,6 +65,40 @@ class IntegrationTestSamplesExecutor extends CommandExecutor {
 
     @Override
     protected int run(String executable, List<String> args, List<String> flags, OutputStream outputStream) {
+        try {
+            GradleExecuter executer = createExecuter(args, flags);
+            if (expectFailure) {
+                // TODO(mlopatkin) sometimes it is still possible to save the configuration cache state in the event of failure.
+                //  We need to figure out how to separate expected failure from the CC store failure.
+                ExecutionFailure result = executer.runWithFailure();
+                outputStream.write((result.getOutput() + result.getError()).getBytes());
+            } else {
+                ExecutionResult result = executer.run();
+                outputStream.write(result.getOutput().getBytes());
+
+                if (shouldCheckLoadingFromConfigurationCache()) {
+                    rerunReusingStoredConfigurationCacheEntry(args, flags);
+                }
+            }
+            return expectFailure ? 1 : 0;
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void rerunReusingStoredConfigurationCacheEntry(List<String> args, List<String> flags) {
+        // Rerun tasks. If the task is up-to-date, then its actions aren't executed.
+        // This might hide issues with Groovy closures, as the methods referenced inside the closure are resolved dynamically.
+        try {
+            createExecuter(args, flags).withArgument("--rerun-tasks").run();
+        } catch (UnexpectedBuildFailure e) {
+            UnexpectedBuildFailure tweakedException = new UnexpectedBuildFailure("Failed to rerun the build when reusing the configuration cache entry.\n" + e.getMessage());
```

Comment:
A little trick to provide a better error message, while printing the single original stack. A typical causal chain is not very suitable, as it either pushes the interesting Gradle output down to `caused by: ...` section or forces us to print the output twice. With this, we can have output like [there](https://ge.gradle.org/s/koesyiw2jzcv6/tests/:docs:docsTest/org.gradle.docs.samples.Bucket2SnippetsTest/snippet-tasks-incremental-task_groovy_incrementalTaskChangedProperty.sample?top-execution=1). Adding the original exception as a cause actually [looks acceptable in GE](https://ge.gradle.org/s/cncnkjnb5xvla/tests/:docs:docsTest/org.gradle.docs.samples.Bucket2SnippetsTest/snippet-tasks-incremental-task_groovy_incrementalTaskChangedProperty.sample?top-execution=1), but the stacktrace would be unfolded everywhere else.